File: examples/go/flaky_test.go
package flaky

import (
	"math/rand"
	"os"
	"strconv"
	"testing"
	"time"
)

// Initialize random seed from GO_TEST_SEED environment variable
func init() {
	seed := int64(42) // default seed
	if seedStr := os.Getenv("GO_TEST_SEED"); seedStr != "" {
		if parsedSeed, err := strconv.ParseInt(seedStr, 10, 64); err == nil {
			seed = parsedSeed
		}
	}
	rand.Seed(seed)
}

// TestRandomFailure demonstrates a test that fails randomly (~30% of the time)
// This simulates race conditions or non-deterministic behavior
func TestRandomFailure(t *testing.T) {
	value := rand.Float64()

	// Fails when value > 0.7
	if value > 0.7 {
		t.Errorf("Random failure: got %.3f, expected <= 0.7", value)
	}
}

// TestTimingDependent demonstrates a test that depends on timing
// This simulates timeout issues or performance-dependent tests
func TestTimingDependent(t *testing.T) {
	// Simulate variable processing time
	delay := time.Duration(rand.Intn(5)+1) * time.Millisecond
	time.Sleep(delay)

	// Fails if processing takes "too long" (> 4ms)
	if delay > 4*time.Millisecond {
		t.Errorf("Operation too slow: %v", delay)
	}
}

// TestOrderDependency demonstrates a test that depends on execution order
// This simulates shared state issues
func TestOrderDependency(t *testing.T) {
	var items []string

	// Simulate checking a cache that may or may not have items
	if rand.Float64() > 0.5 {
		items = append(items, "existing_item")
	}

	// Fails when cache is unexpectedly populated
	if len(items) != 0 {
		t.Errorf("Expected empty cache, found %d items", len(items))
	}
}

// TestBoundaryCondition demonstrates a test at boundary conditions
// This simulates off-by-one errors
func TestBoundaryCondition(t *testing.T) {
	// Simulate calculating a threshold
	calculatedValue := rand.Intn(5) + 98 // Range: 98-102
	threshold := 100

	// Fails when value exceeds threshold
	if calculatedValue > threshold {
		t.Errorf("Value %d exceeds threshold %d", calculatedValue, threshold)
	}
}

// TestConcurrentAccess demonstrates concurrent access patterns
// This simulates race conditions with shared resources
func TestConcurrentAccess(t *testing.T) {
	// Simulate checking if resource is locked
	isLocked := rand.Float64() > 0.5

	// Fails when resource is locked
	if isLocked {
		t.Error("Resource is locked by another process")
	}
}

// TestNetworkSimulation demonstrates network flakiness
// This simulates unreliable network conditions
func TestNetworkSimulation(t *testing.T) {
	// Simulate network response success rate
	successRate := rand.Float64()

	// Fails 20% of the time (simulating network issues)
	if successRate <= 0.2 {
		t.Errorf("Network request failed: %.3f", successRate)
	}
}

// TestMapIteration demonstrates non-deterministic map iteration
// Go maps have random iteration order
func TestMapIteration(t *testing.T) {
	m := map[string]int{
		"a": 1,
		"b": 2,
		"c": 3,
	}

	// Get first key (non-deterministic in Go)
	var firstKey string
	for k := range m {
		firstKey = k
		break
	}

	// This test is intentionally flaky - map iteration order is random
	// But with seeded random, we can make it more predictable
	expectedKeys := []string{"a", "b", "c"}
	expected := expectedKeys[rand.Intn(len(expectedKeys))]

	if firstKey != expected {
		t.Errorf("Expected first key to be %s, got %s", expected, firstKey)
	}
}

// TestChannelRace demonstrates channel race conditions
// This simulates timing issues with goroutines
func TestChannelRace(t *testing.T) {
	ch := make(chan int, 1)

	// Randomly decide to send or not
	if rand.Float64() > 0.5 {
		ch <- 1
	}

	// Try to receive (may block or succeed)
	select {
	case val := <-ch:
		if val != 1 {
			t.Errorf("Unexpected value: %d", val)
		}
	case <-time.After(1 * time.Millisecond):
		t.Error("Channel receive timeout - no value sent")
	}
}

File: examples/typescript-jest/flaky.test.ts
/**
 * Example flaky tests for TypeScript/Jest
 *
 * These tests demonstrate various flaky patterns that can occur in real applications.
 * They use the JEST_SEED environment variable for reproducible randomness.
 */

describe('Flaky Tests - TypeScript/Jest', () => {

  describe('Random Failure', () => {
    test('should fail randomly ~30% of the time', () => {
      const value = Math.random();

      // Fails when value > 0.7
      expect(value).toBeLessThanOrEqual(0.7);
    });
  });

  describe('Timing Dependent', () => {
    test('should fail when operation takes too long', async () => {
      // Simulate variable processing time
      const delay = Math.floor(Math.random() * 5) + 1;
      await new Promise(resolve => setTimeout(resolve, delay));

      // Fails if processing takes "too long" (> 4ms)
      expect(delay).toBeLessThan(4);
    });
  });

  describe('Order Dependency', () => {
    test('should fail based on execution order', () => {
      const items: string[] = [];

      // Simulate checking a cache that may or may not have items
      if (Math.random() > 0.5) {
        items.push('existing_item');
      }

      // Fails when cache is unexpectedly populated
      expect(items).toHaveLength(0);
    });
  });

  describe('Boundary Condition', () => {
    test('should fail at edge cases', () => {
      // Simulate calculating a threshold
      const calculatedValue = Math.floor(Math.random() * 5) + 98; // Range: 98-102
      const threshold = 100;

      // Fails when value exceeds threshold
      expect(calculatedValue).toBeLessThanOrEqual(threshold);
    });
  });

  describe('Concurrent Access', () => {
    test('should fail on simulated race condition', () => {
      // Simulate checking if resource is locked
      const isLocked = Math.random() > 0.5;

      // Fails when resource is locked
      expect(isLocked).toBe(false);
    });
  });

  describe('Network Simulation', () => {
    test('should fail on simulated network issue', () => {
      // Simulate network response success rate
      const successRate = Math.random();

      // Fails 20% of the time (simulating network issues)
      expect(successRate).toBeGreaterThan(0.2);
    });
  });

  describe('Async Operation', () => {
    test('should fail on async timing issues', async () => {
      // Simulate async operation that may or may not complete in time
      const willTimeout = Math.random() > 0.7;

      const promise = new Promise<string>((resolve, reject) => {
        const delay = willTimeout ? 150 : 50;
        setTimeout(() => resolve('done'), delay);
      });

      // Fails when operation times out
      await expect(promise).resolves.toBe('done');

      // Additional assertion that makes it flaky
      expect(willTimeout).toBe(false);
    });
  });

  describe('Promise Race', () => {
    test('should fail on promise race condition', async () => {
      const delay1 = Math.random() * 100;
      const delay2 = Math.random() * 100;

      const promise1 = new Promise(resolve =>
        setTimeout(() => resolve('first'), delay1)
      );
      const promise2 = new Promise(resolve =>
        setTimeout(() => resolve('second'), delay2)
      );

      const winner = await Promise.race([promise1, promise2]);

      // Flaky: depends on which promise resolves first
      expect(winner).toBe('first');
    });
  });

  describe('Array Randomization', () => {
    test('should fail on random array operations', () => {
      const items = [1, 2, 3, 4, 5];

      // Shuffle array randomly
      for (let i = items.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [items[i], items[j]] = [items[j], items[i]];
      }

      // Flaky: expects specific order after random shuffle
      expect(items[0]).toBe(1);
    });
  });

  describe('Mock Timing', () => {
    test('should fail on mock timing issues', () => {
      const callback = jest.fn();

      // Randomly call callback
      if (Math.random() > 0.5) {
        callback();
      }

      // Fails when callback not called
      expect(callback).toHaveBeenCalled();
    });
  });
});
